static/js/public/manage-members.js

```diff
@@ -1,194 +1,247 @@
+  );
+}
 
-      if (!checkbox.checked && originalMember.roles.includes(role)) {
-        stateBox.classList.add("remove");
-        stateBox.classList.remove("add");
-      }
+function getChangeCount(member, members) {
+  const oldMember = members.find((data) => data.id === member.id);
 
-      if (
-        (checkbox.checked && originalMember.roles.includes(role)) ||
-        (!checkbox.checked && !originalMember.roles.includes(role))
-      ) {
-        stateBox.classList.remove("add", "remove");
-      }
+  let numberOfChanges = 0;
 
-      if (
-        JSON.stringify(member.roles.sort()) !==
-        JSON.stringify(originalMember.roles.sort())
-      ) {
-        member.dirty = true;
-      } else {
-        member.dirty = false;
-      }
+  if (member.roles.length >= oldMember.roles.length) {
+    numberOfChanges = member.roles.filter((x) => {
+      return oldMember.roles.indexOf(x) === -1;
+    }).length;
+  } else {
+    numberOfChanges = oldMember.roles.filter((x) => {
+      return member.roles.indexOf(x) === -1;
+    }).length;
+  }
 
-      const dirtyState = newData.filter((data) => data.dirty);
+  return numberOfChanges;
+}
 
-      let changeCount = 0;
+function getChangesToggleText(changeCount) {
+  if (!changeCount) {
+    return "No changes";
+  } else if (changeCount === 1) {
+    return "1 change";
+  } else {
+    return `${changeCount} changes`;
+  }
+}
 
-      dirtyState.forEach((member) => {
-        const oldMember = membersData.find((data) => data.id === member.id);
+function buildTable(updatedMembers, STATE) {
+  const tbody = document.querySelector("[data-js-members-table-body]");
+  const template = document.querySelector(
+    "[data-js-members-table-row-template]"
+  );
 
-        let numberOfChanges = 0;
+  tbody.innerHTML = "";
 
-        if (member.roles.length >= oldMember.roles.length) {
-          numberOfChanges = member.roles.filter((x) => {
-            return oldMember.roles.indexOf(x) === -1;
-          }).length;
-        } else {
-          numberOfChanges = oldMember.roles.filter((x) => {
-            return member.roles.indexOf(x) === -1;
-          }).length;
-        }
+  updatedMembers.forEach((member) => {
+    const clone = template.content.cloneNode(true);
+    const memberDisplayNameField = clone.querySelector(
+      "[data-js-member-display-name]"
+    );
+    const memberEmailField = clone.querySelector("[data-js-member-email]");
 
-        changeCount += numberOfChanges;
-      });
+    memberDisplayNameField.innerText = member.displayname;
+    memberEmailField.innerText = member.email;
 
-      if (dirtyState.length) {
-        if (changeCount === 1) {
-          saveButton.innerText = "Save 1 change";
-        } else {
-          saveButton.innerText = `Save ${changeCount} changes`;
-        }
-      } else {
-        saveButton.innerText = "Save changes";
+    const roleCheckboxes = clone.querySelectorAll("[data-js-role-checkbox]");
+    roleCheckboxes.forEach((checkbox) => {
```

Comment:
You should do:
```suggestion
    Array.from(roleCheckboxes).forEach((checkbox) => {
```

or something similar, so you don't do `forEach` on DOM nodes.
This comment applies to other places where you use `queryselectorAll`.